### ISSUE TITLE
sum method for viennacl::matrix object

### ISSUE THREAD

**Issue by cdeterman** [viennacl/viennacl-dev#211]
There doesn't appear to be a method for taking the sum of all `viennacl::matrix` object.  There is `viennacl::linalg::sum` for `viennacl::vector` objects and there are the `row_sum`/`col_sum` methods.  Am I correct that this method does not exist?

For example, the following does not work:

```c++
#include <iostream>

// Use OpenCL with ViennaCL
#define VIENNACL_WITH_OPENCL 1

#include "viennacl/ocl/backend.hpp"
#include "viennacl/matrix.hpp"
#include "viennacl/linalg/sum.hpp"

int main(){
    viennacl::context ctx;
    ctx = viennacl::context(viennacl::ocl::get_context(static_cast<long>(0)));

    float s = 3;
    viennacl::matrix<float> pA; 
    pA = viennacl::scalar_matrix<float>(4, 4, s, ctx);

    float out = viennacl::linalg::sum(pA);
    std::cout << out << std::endl;
}
```

A quick workaround for this is to loop over all rows or columns but that seems wrong in this case.  As far as I understand, the matrices are essentially vector objects in GPU memory anyway.  This should be extensible to `matrix` objects.

**Comment by karlrupp** [comment]
Have you tried 

    viennacl::linalg::sum(viennacl::linalg::row_sum(pA));

? Adding a convenience overload for matrices is a good idea, yes.

**Comment by cdeterman** [comment]
@karlrupp that does work, thanks very much.  A convenience overload is good but I am also wondering about the internals.  I wonder if it is most performant to basically have two `sum` calls here versus having the internal OpenCL kernel do it all in one go.  Not sure how involved that would be or if would be worth the effort.  Any thoughts?

**Comment by karlrupp** [comment]
Yes, a single kernel is a bit faster, yes, but there may not be a single application where this is really important (i.e. a performance bottleneck).

**Comment by tsmithe** [comment]
Hi Karli,

I noticed that row_sum creates a vector of ones and then does a prod of the matrix with that vector. Doesn't this incur some overhead, in the creation of the all_ones vector, and then the unneeded multiplications of the matrix elements by 1?

**Comment by karlrupp** [comment]
Yes, it incurs overhead. I'm happy to change/optimize it if somebody reports it as a performance bottleneck. For the time being, it's maintenance-optimized :-)

**Comment by tsmithe** [comment]
Thanks Karli. By performance bottleneck, do you mean 'the complexity of this operation dominates a given computation'?

I do think -- as a wish-list item -- it would be nice to see ViennaCL have more capabilities for broadcasting and reduction (à la NumPy). But perhaps that's more ArrayFire's niche?
